src/TBuffer.cpp

```diff
@@ -605,20 +1156,52 @@ inline int TBuffer::lookupColor(const QString & s, int pos )
     return -1; // unbeendete sequenz
 }
 
-
-void TBuffer::translateToPlainText( std::string & s )
+void TBuffer::translateToPlainText( std::string & incoming, const bool isFromServer )
 {
-    //cout << "TRANSLATE<"<<s<<">"<<endl;
+    // As well as enabling the prepending of left-over bytes from last packet
+    // from the MUD server this may help in high frequency interactions to
+    // protect this process from the supplied string being modified
+    // asynchronously by the QNetwork code that runs in another thread:
+    std::string localBuffer;
+
+    Host * pHost = mpHost;
+    if (! pHost) {
+        qWarning() << "TBuffer::translateToPlainText(...) ERROR: Cannot access Host instance at this time - data has been lost.";
+        return; // We really have a problem
+    }
+
+    if (isFromServer
+        && pHost->mTelnet.getEncoding() == QLatin1String("UTF-8")) {
+
+        localBuffer = mIncompleteUtf8SequenceBytes + incoming;
+        mIncompleteUtf8SequenceBytes.clear();
+    } else {
+        localBuffer = incoming;
+    }
+
+    const QVector<QChar> encodingLookupTable = csmEncodingTable.value(pHost->mTelnet.getEncoding());
+    // If the encoding is "ASCII", "ISO 8859-1" or "UTF-8" (which are not in the
+    // table) encodingLookupTable will be empty otherwise the 128 values in the
+    // returned table will be used for all the text data that gets through the
+    // following ANSI code and other out-of-band data processing - doing this
+    // mean that a (fast) lookup in the QVector can be done as opposed to a
+    // a repeated switch(...) and branch to one of a series of decoding methods
+    // each with another up to 128 value switch()
+
     speedAppend = 0;
     speedTP = 0;
     int numCodes=0;
     speedSequencer = 0;
     mUntriggered = lineBuffer.size()-1;
-    msLength = s.length();
+    msLength = incoming.length();
```

Comment:
Oh, blast, if the odd byte (or three) is carried over this will be off - it should be `localbuffer.length()` - OTOH I do not image that `mIncompleteUt8SequenceBytes` gets used _that_ often.